### PR TITLE
Rework DRAM microbenchmark to be much simpler

### PR DESCRIPTION
**Pull request by psalz**
This reworks the DRAM microbenchmark as discussed in #25.

I've decided to interpret the `--size` parameter as the buffer size to be copied, in MiB. This makes it easier to compare copying `float`s and `double`s, as otherwise the benchmark would always cap out once the memory limit is reached for the latter.

I also got rid of the "pattern_" prefix in the filename, as discussed via email.

I wanted to also add 2D/3D copies, but mapping the single `--size` parameter to nice, equally sized ranges in 2D and especially 3D is non-trivial - I'll look into it on Monday. Also explicit 3D copies are not yet supported in hipSYCL when targeting CPU (although it [looks like](https://github.com/ROCm-Developer-Tools/HIP/blob/master/docs/markdown/CUDA_Runtime_API_functions_supported_by_HIP.md) HIP might support it now) - is it supported in hipCPU @illuhad?

### CODE REVIEW

**Comment by illuhad** [comment]
> but mapping the single --size parameter to nice, equally sized ranges in 2D and especially 3D is non-trivial

Couldn't we just say:
* 1D copies `(size*size)` data elements
* 2D copies `(size, size)` data elements
* 3D copies `(<constant>, size, size)` data elements

for some relatively small constant in the 3D case?

No, I don't think it's in hipCPU. However, I don't think we need it necessarily here:
* we could just submit an empty dummy kernel, which will also cause data to migrated to device
* On CPU, hipSYCL follows a unified memory model, i.e. no data transfer between host and "device" will happen anyway *if* the buffer was constructed owning the initial data pointer.

File: micro/DRAM.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        output_buf(s::range<1>(buffer_size)) {}
+
+  void setup() {
+    input.resize(buffer_size, 33.f);
+    args.device_queue
+        .submit([&](cl::sycl::handler& cgh) {
+          auto acc = input_buf.template get_access<s::access::mode::discard_write>(cgh);
+          cgh.copy(input.data(), acc);
+        })
+        .wait();
```
Comment (illuhad) [inline micro/DRAM.cpp]: I would suggest to move this `wait()` to the benchmark framework in `common.h` before it executes `run()`. I suspect other benchmarks may forget that synchronization is necessary here, so it's better to ensure globally that before `run()` is executed, the queue is empty.

**Comment by psalz** [comment]
> Couldn't we just say:
> 
>  * 1D copies `(size*size)` data elements
>  * 2D copies `(size, size)` data elements
>  * 3D copies `(<constant>, size, size)` data elements
> 
> for some relatively small constant in the 3D case?

I think we should at least make sure that the amount of data being copied is the same in all three variants, so doing e.g. `(size^3), (size^2, size), (size, size, size)`. I'd also like to normalize based on the data type, so maybe `(size^3 / sizeof(T)), (size^2 / sizeof(T), size), (size / sizeof(T), size, size)` ..?

> No, I don't think it's in hipCPU. However, I don't think we need it necessarily here:
> 
> * we could just submit an empty dummy kernel, which will also cause data to be migrated to device
> * On CPU, hipSYCL follows a unified memory model, i.e. no data transfer between host and "device" will happen anyway _if_ the buffer was constructed owning the initial data pointer.

Okay, I already feared we'd have to go for the dummy kernel ;-). In that case I'd however prefer to not repeat this hack all over the place (as many other benchmarks still need to be converted to eagerly copy data), but instead have some helper function to do it. I can see this being done either with a free function, or by adding a method to `BenchmarkArgs`. In the latter case, I would propose to rename the args to `BenchmarkContext` -- calling it "args" is a stretch already anyways (as it has the queue, result consumer, etc).

While we're at it, I would also like to have a `log()` function somewhere (on the context..?) that allows me to print stuff like the actual 2/3D kernel size to the console, if e.g. a `--verbose` parameter is set.

File: micro/DRAM.cpp (hunk 1 above)
Comment (psalz) [reply micro/DRAM.cpp]: Alright, will do!

**Comment by psalz** [comment]
I've implemented the changes as discussed. The benchmark now also supports 2D and 3D buffers (and I'm getting some interesting results - likely due to suboptimal work group sizes).

I'm now using the dummy kernel instead of explicit memory operations to force-initialize the input buffer, but I'm not very happy with this approach. The fact that the copy will only be performed for the first run leaves a bad taste in my mouth, even though buffers should be large enough that caching effects are negligible. Ideally we'd have completely new buffers for each run of the benchmark. We could do this by e.g. wrapping them in `unique_ptr`, but that's also not very nice. I have another idea how we can avoid state in benchmarks altogether, and I might open an RFC PR on that later today.

Since I'm making use of `if constexpr`, I've also snuck in a commit to raise the C++ standard to 17. I think that's reasonable for this project, as hipSYCL and ComputeCpp now both use Clang 8 (or higher).

**Comment by psalz** [comment]
Scratch my previous point, I forgot that the benchmark class is being newly instantiated for every run. Besides not being super-pretty, I think this implementation should be fine now!

**Pull request merged**